### ISSUE TITLE
Add option to ignore guard clauses in ReturnCount rule

### ISSUE THREAD

**Issue by 3flex** [arturbosch/detekt#1868]
## Expected Behavior of the rule
ReturnCount rule has an option to ignore use of [guard clauses](https://www.refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html).

## Context
In some functions it is cleaner to have a series of guard clauses at the top of the function to return early than some other methods of achieving the same result.

An option to ignore guard clauses can be added to the rule. When enabled the logic could be to ignore `return`s for all statements at the beginning of a function, but only when statements are in one of the following forms:

```kotlin
if (expression) return
val x = person.name ?: return
```

These would have to be the first declarations in the function body. As soon as a statement not in that form appeared then ReturnCount would start counting instances of return.

To be debated is whether or not `require` or `check` functions could appear before guard clauses. They serve a very similar function. IMHO these functions (and variants) could appear before guard clauses as well, since it would be better for the function to throw an exception due to unexpected/incorrect state before guard clauses are checked, as guard clauses are part of normal program flow.

**Comment by amitdash291** [comment]
Picking this feature up.
Targeting to support `if` condition guard clauses (need not be the first statement necessarily).
Planning to add the configuration `excludeGuardClauses` for this purpose with a default value of `false`.

**Comment by 3flex** [comment]
Thanks @amitdash291! I'd still like to see the guard clauses be ignored when they're the first statements (or following `require` or `check` statements) as returns from other areas of the function, even as part of a simple if statement, couldn't really be considered "guards", instead they're just part of the function's logic.

Also `ignoreGuardClauses` would be better as the config option name I think.

Looking forward to the PR!

**Comment by amitdash291** [comment]
> Thanks @amitdash291! I'd still like to see the guard clauses be ignored when they're the first statements (or following `require` or `check` statements) as returns from other areas of the function, even as part of a simple if statement, couldn't really be considered "guards", instead they're just part of the function's logic.
> 
> Also `ignoreGuardClauses` would be better as the config option name I think.
> 
> Looking forward to the PR!

@3flex, thanks for the response.

Support for the Elvis operator guard clauses (eg: `val x = person.name ?: return`) is also possible. The only remaining piece of the puzzle is to make sure the clauses are the first statement of the functions. I am working on it.

**Comment by amitdash291** [comment]
@3flex, the PR is ready and approved. It would be great if you could review it too :)

**Comment by amitdash291** [comment]
Currently, the PR doesn't support excluding return count for _guard clauses followed by require and check functions_. I had a few questions to be clarified before the same can be done with corresponding test cases.

For eg: does the following cases also qualify as exclusion scenarios:
1. ```kotlin
fun testFun(testVal: String?) {
  requireThat(testVal != null)
  checkIf(testVal != null)
  if (testVal != null) return
}
 ```

2. ```kotlin
fun testFun(testVal: String?) {
  require(testVal != null)
  checkValidations(testVal != null)
  if (testVal != null) return
}
 ```
Or should we watch out for only `check` and `require` strings in the function name?

I would recommend doing this in the next iteration, possibly as a part of a follow-up issue, where we can get these questions clarified. I would be happy to pick it up.

**Comment by schalkms** [comment]
@amitdash291 thanks for implementing this feature. Could you please formulate the missing improvements in a new issue? Otherwise, it becomes hard to track the details here.

**Comment by amitdash291** [comment]
@schalkms thanks a lot for merging the PR!
I will go ahead and copy the remaining improvements in a new issue soon.
